cola/widgets/toolbar.py

```diff
@@ -0,0 +1,670 @@
+        self.widget = widget
+        self.Base = Base
+
+        widget.setAcceptDrops(True)
+        widget.setSelectionMode(widget.SingleSelection)
+        widget.setDragEnabled(True)
+        widget.setDropIndicatorShown(True)
+
+    def dragEnterEvent(self, event):
+        widget = self.widget
+        self.Base.dragEnterEvent(widget, event)
+
+    def dragMoveEvent(self, event):
+        widget = self.widget
+        self.Base.dragMoveEvent(widget, event)
+
+    def dragLeaveEvent(self, event):
+        widget = self.widget
+        self.Base.dragLeaveEvent(widget, event)
+
+    def dropEvent(self, event):
+        widget = self.widget
+        event.setDropAction(Qt.MoveAction)
+        self.Base.dropEvent(widget, event)
+
+    def get_items(self):
+        widget = self.widget
+        base = self.Base
+        items = [base.item(widget, i) for i in range(base.count(widget))]
+
+        return items
+
+
+class DraggableListWidget(QtWidgets.QListWidget):
+    Mixin = DraggableListMixin
+
+    def __init__(self, parent=None):
+        QtWidgets.QListWidget.__init__(self, parent)
+
+        self.setAcceptDrops(True)
+        self.setSelectionMode(self.SingleSelection)
+        self.setDragEnabled(True)
+        self.setDropIndicatorShown(True)
+
+        self._mixin = self.Mixin(self, QtWidgets.QListWidget)
+
+    def dragEnterEvent(self, event):
+        return self._mixin.dragEnterEvent(event)
+
+    def dragMoveEvent(self, event):
+        return self._mixin.dragMoveEvent(event)
+
+    def dropEvent(self, event):
+        return self._mixin.dropEvent(event)
+
+    def add_separator(self, title, data):
+        item = QtWidgets.QListWidgetItem()
+        item.setText(title)
+        item.setData(Qt.UserRole, data)
+
+        self.addItem(item)
+
+    def add_item(self, title, data, icon_text=None):
+        item = QtWidgets.QListWidgetItem()
+        item.setText(N_(title))
+        item.setData(Qt.UserRole, data)
+
+        if icon_text is not None:
+            icon = getattr(icons, icon_text, None)
+            item.setIcon(icon())
+
+        self.addItem(item)
+
+    def get_items(self):
+        return self._mixin.get_items()
+
+
+class ToolbarTreeWidget(standard.TreeView):
+    def __init__(self, parent):
+        standard.TreeView.__init__(self, parent)
+
+        self.setDragEnabled(True)
+        self.setDragDropMode(QtWidgets.QAbstractItemView.DragOnly)
+        self.setSelectionMode(QtWidgets.QAbstractItemView.SingleSelection)
+        self.setDropIndicatorShown(True)
+        self.setRootIsDecorated(True)
+        self.setHeaderHidden(True)
+        self.setAlternatingRowColors(False)
+        self.setSortingEnabled(False)
+
+        self.setModel(QtGui.QStandardItemModel())
+
+    def create_item(self, name, data):
+        item = QtGui.QStandardItem()
+
+        item.setEditable(False)
+        item.setDragEnabled(True)
+        item.setText(N_(name))
+        item.setToolTip(N_(name))
```

Comment:
I think the same. I will work on the refactoring of the tree and try to reproduce the issues.